### PR TITLE
Fixes #1754

### LINKED ISSUE #1754: Typing error on database.dump()
## Case
Bug

## Issue
Typescript does not know database.dump() is a promise:

## Info
  - Environment: (Node.js/browser/electron/etc..)
  - Adapter: (IndexedDB)
  - Stack: (Typescript, Angular)

## Code

In rx-database.ts:
```
/**
* export to json
*/
dump(_decrypted: boolean = false, _collections?: string[]): string[] | null {
    throw pluginMissing('json-dump');
}
```

In json-dump.ts:
```
function dumpRxDatabase(
    this: RxDatabase,
    decrypted = false,
    collections?: string[]
): Promise<any> {
    const json: any = {
        name: this.name,
        instanceToken: this.token,
        encrypted: false,
        passwordHash: null,
        collections: []
    };

    if (this.password) {
        json.passwordHash = hash(this.password);
        if (decrypted) json.encrypted = false;
        else json.encrypted = true;
    }

    const useCollections = Object.keys(this.collections)
        .filter(colName => !collections || collections.includes(colName))
        .filter(colName => colName.charAt(0) !== '_')
        .map(colName => this.collections[colName]);

    return Promise.all(
        useCollections
            .map(col => col.dump(decrypted))
    ).then(cols => {
        json.collections = cols;
        return json;
    });
}
```
In the generated rx-database.d.ts:
```
    /**
     * export to json
     */
    dump(_decrypted?: boolean, _collections?: string[]): string[] | null;

```

### PR DESCRIPTION
**Pull request by PVermeer**
#1755  This PR contains:
 - IMPROVED typings
 - A BUGFIX

## Describe the problem you have without this PR
#1754 

**Improved import export types:**
They are now depending on the encryption status via overloads. Logic:
When _decrypted parameter on x.dump() is set to false or or omitted, it will set all base properties from the RxDocumentType to any (we don't know if it's encrypted or not). When true it will be typed as the RxDocumentType (we definitely know values will not be encrypted).

I think this the best solution without restructuring the code thus making big changes.

**Added types on import-export tests and added negative tests to the database import tests:**
It is difficult in this code structure to let TypeScript infer most types. So I added some types to properties used in tests for type testing purposes.

**Added types to RxJsonSchema's and fixed some issues with them:**
I basically strong typed the test schema's so errors are visible now. Also fixed the some typ errors on them.

### CODE REVIEW

File: src/rx-database.ts [hunk 1]
```diff
@@ -330,7 +333,7 @@ export class RxDatabaseBase<Collections = CollectionsOfDatabase> {
     /**
      * delete all data of the collection and its previous versions
      */
-    removeCollection(collectionName: string): Promise<string[]> {
+    removeCollection(collectionName: string): Promise<void> {
```
Comment (PVermeer) [inline src/rx-database.ts]: Changed because of type errors now that this.lockedRun returns the function type.

File: src/types/rx-collection.ts [hunk 2]
```diff
@@ -19,7 +19,7 @@ export interface RxCollectionCreator {
     name: string;
     schema: RxJsonSchema;
     pouchSettings?: PouchSettings;
-    migrationStrategies?: NumberFunctionMap;
+    migrationStrategies?: KeyFunctionMap;
```
Comment (PVermeer) [inline src/types/rx-collection.ts]: This is the type as defined in the class RxCollectionBase that is using this property.

**Comment by pubkey** [comment]
I restarted the CI so it is green now.
I noticed that this PR increases the build-size by 400 byte. See [here](https://travis-ci.org/pubkey/rxdb/jobs/624557679?utm_medium=notification&utm_source=github_status) compared to [here](https://travis-ci.org/pubkey/rxdb/jobs/626207941?utm_medium=notification&utm_source=github_status)
Can you run a build and make a diff on the dist-folder to check what increased that much?

**Comment by PVermeer** [comment]
Sorry don't have time anymore. But you're reply from text to byte is already 215 bytes and gZipped 170 bytes. I added some types + some text to go with it, so seems about right.

**Comment by pubkey** [comment]
I fixed the build size. The problem was the unnecessary usage of async at one point. This made babel using the regenernator runtime.
```typescript
         return _this3._spawnPouchDB(collectionName, v);
       });
     }) // remove documents
-    .then(function (pouches) {
-      return Promise.all(pouches.map(function (pouch) {
-        return _this3.lockedRun(function () {
-          return pouch.destroy();
-        });
+    .then(
+    /*#__PURE__*/
+    function () {
+      var _ref = _asyncToGenerator(
+      /*#__PURE__*/
+      _regeneratorRuntime.mark(function _callee(pouches) {
+        return _regeneratorRuntime.wrap(function _callee$(_context) {
+          while (1) {
+            switch (_context.prev = _context.next) {
+              case 0:
+                _context.next = 2;
+                return Promise.all(pouches.map(function (pouch) {
+                  return _this3.lockedRun(function () {
+                    return pouch.destroy();
+                  });
+                }));
+
+              case 2:
+              case "end":
+                return _context.stop();
+            }
+          }
+        }, _callee);
       }));
-    });
+
+      return function (_x) {
+        return _ref.apply(this, arguments);
+      };
+    }());

```

**Comment by pubkey** [comment]
I merged this now. Thank you @PVermeer

**Pull request merged**